DirectProgramming/Fortran/guided_matrix_mul_OpenMP/README.md

```diff
@@ -0,0 +1,433 @@
+3. Build the program.
+   ```
+   ifx /Qxhost 01_mm_CPU_sequential.f90
+   ```
+4. Run the program.
+   ```
+   mm_sequential.exe
+   ```
+
+### Example Sequential Program Output
+
+The output of the sequential program output is:
+
+```
+Matrix size 2600
+PASSED
+```
+
+Note that when compiler option -xhost (/Qxhost) is used, the compiler optimizes the code for the CPU where the compile is done.
+
+## Modify the Program to Use OpenMP Directives (Parallel Version)
+
+In a parallel region of code, the master thread spawns a team of threads as needed according to the OpenMP directives. Parallelism can be added incrementally by the developer until the desired performance is achieved.
+
+<img src="images/workshare.png" alt="Parallelism: Distributes the execution of loop iterations across the threads" width="500"/>
+
+OpenMP directives all start a line with `!$omp`.
+
+A DO loop can be parallelized by simply wrapping the loop with `!$omp parallel do` and `!$omp end parallel do`.
+
+Optional clauses for `!$omp parallel do` include `shared` and `private`. Variables that are declared shared are shared between the threads. In the parallel version of the sample, the parallelism is done at the outer loop where `j` is the loop index. That makes the value of `j` unique with each thread, so there is no opportunity to write to the same memory location (creating a race condition) in the arrays during the computation. The memory locations of the arrays are shared. Variables that are declared private are unique for each thread. By default, according to the OpenMP standard, the index of a DO loop is private. The indices are declared in the example for clarity.
+
+In the sample, the arrays `a`, `b`, and `c` are shared between the threads as is the variable `n`, the size of the matrix. Those are listed in the shared clause. The indices are declared as private, although that is optional.
+
+### Edit the Parallel Program
+
+1. Open the file `01_mm_CPU_sequential.F90` and save it as `mm_CPU_parallel.F90`. 
+2. In `mm_CPU_parallel.F90`, add two `!$omp` directives, wrapping the matrix multiply function with `!$omp parallel do` and `!$omp end parallel do`: 
+
+   ```
+   !$omp parallel do shared(a, b, c, n), private(i, j, k)
+   ! parallel compute matrix multiplication.
+      do j=1,n
+        do i=1,n
+          do k=1,n
+            c(i,j) = c(i,j) + a(i,k) * b(k,j)
+          enddo
+        enddo
+      enddo
+   !$omp end parallel do
+   ```
+
+## Build and Run the Parallel Program
+
+In order for the Intel Fortran Compiler to recognize OpenMP directives, the program must be compiled with the `-qopenmp` (Linux) or `/Qopenmp` (Windows) option. Without this compiler option, the OpenMP directives are ignored and the program will run sequentially. Note that this can be helpful for debugging the program.
+
+### On Linux
+
+1. Change to the sample directory.
+2. Build the program using the `-qopenmp` option.
+   ```
+   ifx -xhost -qopenmp mm_CPUparallel.f90
```

Comment:
@krisdale I think you need to change this from `mm_CPUparallel.f90` to `02_mm_CPU_parallel.f90`